### ISSUE TITLE
Support _NET_WM_DESKTOP correctly

### ISSUE THREAD

**Issue by Airblader** [i3/i3#2153]
We currently have `_NET_WM_DESKTOP` in `_NET_SUPPORTED`. This seems to have been the case since around 2009 (1a0817eb3912f25b75cfe61854920c52c78493d2), although I don't really see the reason for it even in that commit.

As a `git grep NET_WM_DESKTOP` reveals, we don't do *anything* with this atom – we don't respect it upon map requests, we don't keep it updated and we don't respond to client messages, all of which should be the case according to [the spec](http://standards.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472635472).

We discussed before in #1507 and #938 that we don't want to support this atom, so we should also not claim to support it in the first place – unless we have a good reason to. @stapelberg Are you aware of any reasons?

However, I'd also like to propose supporting it instead. At least giving it (another) shot. I think we can have a canonical way to enumerate workspaces (even only-named ones) according to the spec and we do have sticky windows now, so we could respect the special value by floating the window and making it sticky.

I'd only actually do this, though, if we can verify that proper support of this atom makes pagers (like gnome-panel) work correctly, see [this post](https://www.reddit.com/r/i3wm/comments/402e2v/i3_workspaces_munged_together_by_desktop_tools/?ref=share&ref_source=link) and similar requests. Otherwise we shouldn't worry about the effort. @stapelberg Thoughts?

**Comment by i3bot** [comment]
I don’t see a version number. Could you please copy & paste the output of `i3 --version` into this issue?

**Comment by Airblader** [comment]
awesomeWM/awesome also implements this slightly "incorrect":

````
/** Update the client active desktop.
 * This is "wrong" since it can be on several tags, but EWMH has a strict view
 * of desktop system so just take the first tag.
 * \param c The client.
 */
void
ewmh_client_update_desktop(client_t *c) 
{
    int i;

    for(i = 0; i < globalconf.tags.len; i++)
        if(is_client_tagged(c, globalconf.tags.tab[i]))
        {
            xcb_change_property(globalconf.connection, XCB_PROP_MODE_REPLACE,
                                c->window, _NET_WM_DESKTOP, XCB_ATOM_CARDINAL, 32, 1, &i);
            return;
        }
    /* It doesn't have any tags, remove the property */
    xcb_delete_property(globalconf.connection, c->window, _NET_WM_DESKTOP);                                                                                                                                                                                                                                       
}
```

**Comment by Airblader** [comment]
I think we already have a way to number the workspaces since we also need it for  _NET_CURRENT_DESKTOP.

**Comment by stapelberg** [comment]
> We discussed before in #1507 and #938 that we don't want to support this atom, so we should also not claim to support it in the first place – unless we have a good reason to. @stapelberg Are you aware of any reasons?

I’m not aware of any reasons. I think this was an oversight when developing back then.

> I'd only actually do this, though, if we can verify that proper support of this atom makes pagers (like gnome-panel) work correctly, see this post and similar requests. Otherwise we shouldn't worry about the effort. @stapelberg Thoughts?

Sounds good!

**Comment by Airblader** [comment]
I've just tried `xfce4-panel` and set `_NET_WM_DESKTOP` manually on all my windows (`xprop -format _NET_WM_DESKTOP 32c -set _NET_WM_DESKTOP 0`). Setting the atom was enough to make the pager display everything correctly, so I'll convert this to a feature request instead.

**Comment by Airblader** [comment]
Things to take care of for full spec compliance:

[ ] Set the atom when we manage a window.
[ ] Update the atom when a window is moved.
[ ] Upon managing a window, check if the atom is set. If so, place it accordingly. If the atom is set to 0xFFFFFFFF, make it floating and sticky.
[ ] When withdrawing a window, remove the atom.
[ ] Look for client messages and react to them.

**Comment by Airblader** [comment]
Leftovers:

- [ ] The implementation is currently buggy if, for example, workspaces 1 and 3 are open with windows and then workspace 2 is opened. Whenever workspaces are opened, closed or moved, we need to update the hint on all windows.
- [ ] Tests

**Comment by Airblader** [comment]
@stapelberg Here's one more thing I just noticed. For pagers to display the workspaces correctly when moving a window to another workspace, we'd have to render that workspace. Otherwise the pager doesn't know the geometry of the window. If I had to guess I'd say you aren't too happy with that.

I'd still go for _NET_WM_DESKTOP support since it can be used by tools other than pagers as well. Like bars.
